File: 03rat_in_a_maze.cpp
class Solution {
public:
    vector<string> output;
    bool solve(vector<vector<int>> &m, vector<vector<int>> &visited, int r, int c, int row, int col, string move) {
        if (row == r- 1 and col == c - 1) {
            output.push_back(move);
            return true;
        }
        //checking downward
        if (row + 1 < r and visited[row + 1][col] != 1 and m[row + 1][col] == 1) {
            visited[row][col] = 1;
            solve(m, visited, r, c, row + 1, col, move + 'D');
            visited[row][col] = 0;
        }
        //going left
        if (col - 1 >= 0 and visited[row][col - 1] != 1 and m[row][col - 1] == 1) {
            visited[row][col] = 1;
            solve(m, visited, r, c, row, col - 1, move + 'L');
            visited[row][col] = 0;
        }
        //going right
        if (col + 1 < c and visited[row][col + 1] != 1 and m[row][col + 1] == 1) {
            visited[row][col] = 1;
            solve(m, visited, r, c, row, col + 1, move + 'R');
            visited[row][col] = 0;
        }
        //checking upward
        if (row - 1 >= 0 and visited[row - 1][col] != 1 and m[row - 1][col] == 1) {
            visited[row][col] = 1;
            solve(m, visited, r, c, row - 1, col, move + 'U');
            visited[row][col] = 0;
        }
    }
    vector<string> findPath(vector<vector<int>>& m, int n) {
        // Your code goes here
        int r = m.size();
        int c = m[0].size();
        if (m[0][0] == 0){
            return output;
        }
        vector<vector<int>> visited(r, vector<int>(c, 0));
        solve(m, visited, r, c, 0, 0, "");
        return output;
    }
};
//shortcode
class Solution {
public:
    vector<string> output;
    vector<int> row_list;
    vector<int> col_list;
    bool solve(vector<vector<int>>& m, vector<vector<int>>& visited, int n, int row, int col, string move) {
        if (row == n - 1 and col == n - 1) {
            output.push_back(move);
            return true;
        }
        string ways = "DLRU";
        for (int i = 0; i < 4; i++) {
            int next_row = row + row_list[i];
            int next_col = col + col_list[i];
            if (next_row >= 0 and next_col >= 0 and next_row < n and next_col < n and visited[next_row][next_col] != 1
                and m[next_row][next_col] == 1) {
                visited[row][col] = 1;
                solve(m, visited, n, next_row, next_col, move + ways[i]);
                visited[row][col] = 0;
            }
        }
    }
    vector<string> findPath(vector<vector<int>>& m, int n) {
        // Your code goes here
        vector<vector<int>> visited(n, vector<int>(n, 0));
        if (m[0][0] == 0){
            return output;
        }
        row_list = { 1, 0, 0, -1 };
        col_list = { 0, -1, 1, 0 };
        solve(m, visited, n, 0, 0, "");
        return output;
    }
};

File: 03rat_in_a_maze.py
# time: O(4^(n*m)) and space: O(n*m)
class Solution:
    def __init__ (self):
        self.result = []
    def findPath(self, m, n):
        visited = [[0 for i in range(n)]for j in range(n)]
        if m[0][0] == 0:
            return self.result
        self.solve(m, n, visited, 0, 0, "")
        return self.result
    def solve(self, m, n, visited, row, col, move):
        if row == n -1 and col == n -1:
            self.result.append(move)
            return
        # downward
        if row + 1 < n and visited[row + 1][col] != 1 and m[row + 1][col] == 1:
            visited[row][col] = 1
            self.solve(m, n, visited, row + 1, col, move + 'D')
            visited[row][col] = 0
        #left
        if col - 1 >= 0 and visited[row][col -1] != 1 and m[row][col -1] == 1:
            visited[row][col] = 1
            self.solve(m, n, visited, row, col - 1, move + 'L')
            visited[row][col] = 0
        # right
        if col + 1 < n and visited[row][col + 1] != 1 and m[row][col + 1] == 1:
            visited[row][col] = 1
            self.solve(m, n, visited, row, col + 1, move + 'R')
            visited[row][col] = 0
        # upward
        if row - 1 >= 0 and visited[row - 1][col] != 1 and m[row - 1][col] == 1:
            visited[row][col] = 1
            self.solve(m, n, visited, row - 1, col, move + 'U')
            visited[row][col] = 0
# short code
class Solution:
    def __init__(self):
        self.result = []
        self.x_coordinates = [1, 0, 0, -1]
        self.y_coordinates = [0, -1, 1, 0]
        
    def findPath(self, m, n):
        # to keep track of visited cell, because we can't visit a cell twice
        visited = [[0 for i in range(n)]for j in range(n)]
        if m[0][0] == 0:
            return self.result
        self.solve(m, n, visited, 0, 0, "")
        return self.result

    def solve(self, m, n, visited, curr_row, curr_col, move):
        # we are at the destination
        if curr_row == n - 1 and curr_col == n - 1:
            self.result.append(move)
            return
        # downward
        ways = 'DLRU'
        for ind in range(4):
            next_row = curr_row + self.x_coordinates[ind]
            next_col = curr_col + self.y_coordinates[ind]
            if 0 <= next_row < n and 0 <= next_col < n and visited[next_row][next_col] != 1 and m[next_row][next_col] == 1:
                # do
                visited[curr_row][curr_col] = 1
                # recurse
                self.solve(m, n, visited, next_row, next_col, move + ways[ind])
                # undo
                visited[curr_row][curr_col] = 0
                
